### ISSUE TITLE
feat: auto-sanitise name when creating bot / user

### ISSUE THREAD

**Issue by insertish** [revoltchat/backend#257]
Instead of restricting username value, just set a display name if username is non compliant and copy the sanitised name into username.

**Comment by StaTicDKK17** [comment]
Hello. I would like to help with this. is there already somewhere in the backend where sanitization is happening or do I have to invent that from scratch?

**Comment by insertish** [comment]
Only validation, no sanitisation

https://github.com/revoltchat/backend/blob/main/crates/core/database/src/models/users/model.rs#L187 (create user) uses https://github.com/revoltchat/backend/blob/main/crates/core/database/src/models/users/model.rs#L277 (validate username) which throws an error if the username is invalid

I guess the best way to tackle it is to perform the same steps but instead stripping whatever the blocked elements are from the username, using that as the new username, and setting the display name field to the original username

NB. one needs to check the username is still of minimum length after it is sanitised, one could either replace blocked characters/strings with a placeholder such as `_` or throw an error if it's too short, or a combination or both: strip until we need to append `_` to the end

**Comment by StaTicDKK17** [comment]
So, just to make sure I understand, we currently have some code for validating a username. Currently, if the username contains a specific kind of substring (which looks like protection against sqli), then we currently just throw an error that the username does not accept the criteria. But instead, I should just take the username, and remove the strings that violate the blocked parts. Ensure it is still valid in length, and return the result?

Do you want this action of sanitization to happen inside the validate_username function? Or should it be a different function? Because, then the name of the function doesn't really make great sense anymore since it actually tries to sanitize instead.

**Comment by StaTicDKK17** [comment]
I don't think it makes any sense to try to sanitize for blocked names, since these are supposed to be on the whole username and not only on parts right? Like one could still create a user called something like "revoltFan" or similar, where revolt is a blocked username, but this is only part of the whole username, which contains more?

**Comment by StaTicDKK17** [comment]
Alright, let's say I have created the code to make the feature. How do I contribute, this is my first time contributing to open source :)

Do I open a new branch on the repo and make a PR to main or I have to do something special? I have tried reading the contribution guidelines, but I am not entirely sure if I am right in creating a new branch and creating a PR or what.

**Comment by IAmTomahawkx** [comment]
Youll want to fork the repo, push your changes to your fork on main, and then open a pr from there (github should have a button on your fork)

**Comment by StaTicDKK17** [comment]
Is there any minimum length of a username besides empty? I have searched around in the project for a while but couldn't find anything.

Also, are there tests such that I can test if the functionality works or do I just have to run the dev build and try visually to see if the solution works?

**Comment by insertish** [comment]
Off the top of my head, the request structs have validator properties on them which ensure min length of 2 or whatever it's set to. Maybe this should be put into revolt-config somewhere.

If there are no tests, feel free to write some if you're up for it, there are a few routes/database models with existing tests that you can look at to guide.

Otherwise, you'd need to just follow the development guide in README to test.

**Comment by higgs01** [comment]
I'd love to get working on this one. Is this still up for grabs? and is the desired functionality still the same?

**Comment by ThobiWanKenobi** [comment]
I just want to mention, that I attempted doing it, but it has gone stale due to me being unstable to build and test Revolt backend multiple times. I encourage you to do it if you can get it up and running, I sadly cannot seem to get it up and running, although I have a proposed solution :/

**Comment by higgs01** [comment]
Thanks for the fast response. I'll give it a shot.

**Comment by ThobiWanKenobi** [comment]
Best of luck! Hope you have success with it! :D
